File: supermarket-pricing/typescript/src/Checkout.ts
import type { ComboDeal } from './ComboDeal.js';
import { Money } from './Money.js';
import { Product } from './Product.js';
import { Weight } from './Weight.js';

export class Checkout {
  private readonly products = new Map<string, Product>();
  private readonly quantities = new Map<string, number>();
  private readonly weights = new Map<string, Weight>();
  private readonly comboDeals: ComboDeal[];

  constructor(comboDeals: ComboDeal[] = []) {
    this.comboDeals = [...comboDeals];
  }

  scan(product: Product): void {
    this.ensureRegistered(product);
    this.quantities.set(product.sku, (this.quantities.get(product.sku) ?? 0) + 1);
  }

  scanWeighted(product: Product, weight: Weight): void {
    this.ensureRegistered(product);
    const existing = this.weights.get(product.sku) ?? Weight.zero;
    this.weights.set(product.sku, new Weight(existing.kg + weight.kg));
  }

  total(): Money {
    let result = Money.zero;

    const comboConsumed = new Map<string, number>();

    for (const deal of this.comboDeals) {
      const countA = (this.quantities.get(deal.skuA) ?? 0) - (comboConsumed.get(deal.skuA) ?? 0);
      const countB = (this.quantities.get(deal.skuB) ?? 0) - (comboConsumed.get(deal.skuB) ?? 0);
      const pairs = Math.min(countA, countB);

      if (pairs > 0) {
        result = result.plus(deal.dealPrice.times(pairs));
        comboConsumed.set(deal.skuA, (comboConsumed.get(deal.skuA) ?? 0) + pairs);
        comboConsumed.set(deal.skuB, (comboConsumed.get(deal.skuB) ?? 0) + pairs);
      }
    }

    for (const [sku, product] of this.products) {
      const quantity = (this.quantities.get(sku) ?? 0) - (comboConsumed.get(sku) ?? 0);
      const weight = this.weights.get(sku) ?? Weight.zero;

      if (quantity > 0 || weight.kg > 0) {
        result = result.plus(product.pricingRule.calculate(quantity, weight));
      }
    }

    return result;
  }

  private ensureRegistered(product: Product): void {
    if (!this.products.has(product.sku)) {
      this.products.set(product.sku, product);
    }
  }
}

File: supermarket-pricing/csharp/src/SupermarketPricing/Checkout.cs
namespace SupermarketPricing;

/// <summary>
/// The checkout aggregate. Accumulates scanned items (by count or weight) and
/// computes the total by delegating to each product's pricing rule, then
/// applying any combo deals.
/// </summary>
public sealed class Checkout
{
    private readonly Dictionary<string, Product> _products = new();
    private readonly Dictionary<string, int> _quantities = new();
    private readonly Dictionary<string, Weight> _weights = new();
    private readonly List<ComboDeal> _comboDeals = new();

    public Checkout(IEnumerable<ComboDeal>? comboDeals = null)
    {
        if (comboDeals is not null)
        {
            _comboDeals.AddRange(comboDeals);
        }
    }

    public void Scan(Product product)
    {
        EnsureRegistered(product);
        _quantities[product.Sku] = _quantities.GetValueOrDefault(product.Sku) + 1;
    }

    public void ScanWeighted(Product product, Weight weight)
    {
        EnsureRegistered(product);
        var existing = _weights.GetValueOrDefault(product.Sku);
        _weights[product.Sku] = new Weight(existing.Kg + weight.Kg);
    }

    public Money Total()
    {
        var total = Money.Zero;

        // Track how many items are consumed by combo deals
        var comboConsumed = new Dictionary<string, int>();

        foreach (var deal in _comboDeals)
        {
            var countA = _quantities.GetValueOrDefault(deal.SkuA) - comboConsumed.GetValueOrDefault(deal.SkuA);
            var countB = _quantities.GetValueOrDefault(deal.SkuB) - comboConsumed.GetValueOrDefault(deal.SkuB);
            var pairs = Math.Min(countA, countB);

            if (pairs > 0)
            {
                total += deal.DealPrice * pairs;
                comboConsumed[deal.SkuA] = comboConsumed.GetValueOrDefault(deal.SkuA) + pairs;
                comboConsumed[deal.SkuB] = comboConsumed.GetValueOrDefault(deal.SkuB) + pairs;
            }
        }

        foreach (var (sku, product) in _products)
        {
            var quantity = _quantities.GetValueOrDefault(sku) - comboConsumed.GetValueOrDefault(sku);
            var weight = _weights.GetValueOrDefault(sku);

            if (quantity > 0 || weight.Kg > 0)
            {
                total += product.PricingRule.Calculate(quantity, weight);
            }
        }

        return total;
    }

    private void EnsureRegistered(Product product)
    {
        _products.TryAdd(product.Sku, product);
    }
}
